# Реализация исключений API

Очень распространено реализовать **RESTful API** поверх **Flask**. Одна из первых вещей, с которыми сталкиваются разработчики, - это осознание того, что встроенные исключения недостаточно выразительны для API-интерфейсов и что тип контента `text/html`, который они генерируют, не очень полезен для потребителей API.

Лучшее решение, чем использование **abort** для сигнализации об ошибках при недопустимом использовании API, - это реализовать собственный тип исключения и установить для него обработчик ошибок, который выдает ошибки в формате, ожидаемом пользователем.

## Простой класс исключения

Основная идея состоит в том, чтобы ввести новое исключение, которое может принимать правильное удобочитаемое сообщение, код состояния для ошибки и некоторую дополнительную полезную нагрузку, чтобы дать больше контекста для ошибки.

Это простой пример:

```python
from flask import jsonify

class InvalidUsage(Exception):
    status_code = 400

    def __init__(self, message, status_code=None, payload=None):
        Exception.__init__(self)
        self.message = message
        if status_code is not None:
            self.status_code = status_code
        self.payload = payload

    def to_dict(self):
        rv = dict(self.payload or ())
        rv['message'] = self.message
        return rv
```

Теперь представление может вызвать это исключение с сообщением об ошибке. Кроме того, некоторая дополнительная полезная нагрузка может быть предоставлена в виде словаря через параметр **payload**.

## Регистрация обработчика ошибок

В этот момент представления могут вызвать эту ошибку, но это немедленно приведет к внутренней ошибке сервера. Причина в том, что для этого класса ошибки не зарегистрирован обработчик. Однако это легко добавить:

```python
@app.errorhandler(InvalidUsage)
def handle_invalid_usage(error):
    response = jsonify(error.to_dict())
    response.status_code = error.status_code
    return response
```

## Использование в представлениях

Вот как представление может использовать эту функциональность:

```python
@app.route('/foo')
def get_foo():
    raise InvalidUsage('This view is gone', status_code=410)
```
